### PR TITLE
Refactor sidebar handling to improve performance

### PR DESCRIPTION
**Pull request by jcollins-g**
Since the sidebars are "special", in that they are intended to contain the same contents for symbols defined at the top level of a library or members of a class, take advantage of that and avoid running them through the mustache renderer repeatedly.

This saves about 50% of the time spent in the documentation generation phase on Flutter docs, eliminating ~60,000 extra renderings of various sidebars.   The larger and more complex your package, the bigger of a difference this change makes; it is not really noticeable on trivial sized packages.

** **BREAKING CHANGE** **  This requires the creation (though not use) of `_sidebar_for_library.md` and `_sidebar_for_container.md` for Markdown renderers.  To gain full advantage of the change, custom templates should be modified to use the new `sidebarForLibrary` and `sidebarForContainer` template methods.

### CODE REVIEW

**Comment by jcollins-g** [comment]
@srawlins I think this change should be compatible with your upcoming changes.  Let me know if this looks good in general and I'll finish cleaning it up for submission (probably can further crunch down the sidebar template for Container).

**Comment by srawlins** [comment]
Looks great! I think it will just require re-generation of the templates.mocks.dart file.

File: lib/src/generator/dartdoc_generator_backend.dart [hunk 1]
```diff
@@ -44,14 +44,32 @@ class DartdocGeneratorBackendOptions implements TemplateOptions {
...
 abstract class DartdocGeneratorBackend implements GeneratorBackend {
   final DartdocGeneratorBackendOptions options;
   final Templates templates;
+  final SidebarGenerator<Library> sidebarForLibrary;
+  final SidebarGenerator<Container> sidebarForContainer;
 
   DartdocGeneratorBackend(
       DartdocGeneratorBackendOptions options, this.templates)
-      : options = (options ?? DartdocGeneratorBackendOptions());
+      : options = (options ?? DartdocGeneratorBackendOptions()),
```
Comment (srawlins) [inline lib/src/generator/dartdoc_generator_backend.dart:69]: These parens are unnecessary, right? What about `??=`?

**Review by srawlins — APPROVED** [review]
Very cool!

File: lib/src/generator/dartdoc_generator_backend.dart (hunk 1 above)
Comment (jcollins-g) [reply lib/src/generator/dartdoc_generator_backend.dart:69]: Parentheses are unnecessary, deleted.   `??=` however appears to be illegal in initializer lists.

**Pull request merged**